core/src/test/java/io/questdb/test/griffin/engine/groupby/hyperloglog/HyperLogLogTest.java

```diff
@@ -0,0 +1,264 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2023 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.test.griffin.engine.groupby.hyperloglog;
+
+import io.questdb.cairo.CairoConfiguration;
+import io.questdb.cairo.DefaultCairoConfiguration;
+import io.questdb.griffin.engine.groupby.GroupByAllocator;
+import io.questdb.griffin.engine.groupby.hyperloglog.HyperLogLog;
+import io.questdb.std.Hash;
+import io.questdb.std.Rnd;
+import io.questdb.test.AbstractTest;
+import org.junit.Test;
+
+import static io.questdb.test.griffin.engine.groupby.hyperloglog.HyperLogLogTestUtils.assertCardinality;
+import static io.questdb.test.tools.TestUtils.assertMemoryLeak;
+import static org.junit.Assert.*;
+
+public class HyperLogLogTest extends AbstractTest {
+
+    @Test
+    public void testInvalidPrecision() {
+        assertThrows(IllegalArgumentException.class, () -> new HyperLogLog(3));
+        assertThrows(IllegalArgumentException.class, () -> new HyperLogLog(19));
+    }
+
+    @Test
+    public void testHighCardinality() throws Exception {
```

Comment:
Would be great to add a fuzz test similar to this one, but with randomized inputs and non-constant RNG seeds. See `Unordered8MapTest#testFuzz()` as an example.

Reply:
I like this approach! 

Addressed in 44dffdfcdd14ee8840f23ed23fea7705a3c8cf36. However, there is a small chance that the test I added  will fail. See the comment in `HyperLogLogTestUtils#assertCardinality`. I'm not sure if there is a way around it. I was considering testing the HLL++ from this PR against some other reference implementation. I can research this later.

Reply:
We could relax the assertion with some kind of padding, e.g. 4 standard errors instead of 3. I don't think it's a big deal.